internal/metadataproviders/aws/ec2/chain.go

```diff
@@ -0,0 +1,51 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ec2 // import "github.com/open-telemetry/opentelemetry-collector-contrib/internal/metadataproviders/aws/ec2"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type chainMetadataProvider struct {
+	providers []Provider
+}
+
+func newChainMetadataProvider(providers []Provider) *chainMetadataProvider {
+	return &chainMetadataProvider{providers: providers}
+}
+
+func (p *chainMetadataProvider) ID() string {
+	var providerIDs []string
+	for _, provider := range p.providers {
+		providerIDs = append(providerIDs, provider.ID())
+	}
+	return fmt.Sprintf("Chain [%s]", strings.Join(providerIDs, ","))
+}
+
+func (p *chainMetadataProvider) Get(ctx context.Context) (*Metadata, error) {
+	var errs error
+	for _, provider := range p.providers {
+		if metadata, err := provider.Get(ctx); err != nil {
+			errs = errors.Join(errs, fmt.Errorf("unable to get metadata from %s: %w", provider.ID(), err))
```

Comment:
are all providers expected to have the same metadata? Could lead to a false error if one provider is expected to not have 1 field for example

Reply:
They're all expected to provide the same metadata, which is just everything but the hostname. Hostname has a separate function.